File: 2025/day-1/part-one.ts
const FULL_CIRCLE = 100;

export function partOne(input: Array<string>) {
  let position = 50;
  let result = 0;

  for (const turn of input) {
    const direction = turn[0];
    const value = Number(turn.slice(1));

    if (direction === "L") {
      position -= value;
    } else if (direction === "R") {
      position += value;
    }

    position = position % 100;

    if (position < 0) {
      position += FULL_CIRCLE;
    }
    if (position >= FULL_CIRCLE) {
      position -= FULL_CIRCLE;
    }

    if (position === 0) result += 1;
  }

  return result;
}

File: 2025-rust/day-1/src/part_one.rs
const FULL_CIRCLE: i32 = 100;

pub fn main(input: &Vec<&str>) -> i32 {
    let mut position = 50;
    let mut result = 0;

    for turn in input {
        let direction = turn.chars().next().unwrap();
        let value: i32 = turn[1..].parse().unwrap();

        match direction {
            'L' => position -= value,
            'R' => position += value,
            _ => {}
        }

        position = position.rem_euclid(FULL_CIRCLE);

        if position == 0 {
            result += 1;
        }
    }

    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_simple_right_turn() {
        let input = vec![
            "L68", "L30", "R48", "L5", "R60", "L55", "L1", "L99", "R14", "L82",
        ];

        let result = main(&input);
        assert_eq!(result, 3);
    }
}
